### PR TITLE
[Issue 15271] - Get product variant by id

### LINKED ISSUE #15271: Reaching memory limit when editing product via admin
**Sylius version affected**: 1.2.*

**Description**  

Related to https://github.com/Sylius/Sylius/issues/13742
Reaching memory limit when editing product with multiple variants.
When editing product in Admin, application tries to load all variants at once just to check if there is at least one variant available for product.

**Steps to reproduce**  

Create multiple variants for single product, visit product edit admin page.

**Possible Solution**  

Do single query by id instead of fetching all variants at once.

### PR DESCRIPTION
**Pull request by ulinskas**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.12                   |
| Bug fix?        | yes                                                       |
| New feature?    | no                                                      |
| BC breaks?      | no                                                      |
| Deprecations?   | no |
| Related tickets | fixes #15271                       |
| License         | MIT                                                          |

These changes increase admin performance and optimise memory usage.

### CODE REVIEW

**Comment by jakubtobiasz** [comment]
Hello @ulinskas 👋🏼,
do you have any statistics comparing the performance before and after your changes?🙏🏼

**Comment by ulinskas** [comment]
> Hello @ulinskas 👋🏼, do you have any statistics comparing the performance before and after your changes?🙏🏼

It depends on variant count. 
With few variants you will not notice any difference.
With thousands of variants you will reach memory limit. 
I can give you some numbers if you wish.

**Comment by jakubtobiasz** [comment]
We're aware it'll be noticeable in case of bigger amount of data. It'd be great if you share some numbers 👊🏼.

**Comment by ulinskas** [comment]
@jakubtobiasz 
Here are some simple numbers of variants and performance changes.
I used symfony profiler for simple performance benchmarks.

This is before and after my implementation.
Product with 12 variants:
   controller execution 365.1 ms / 20 MiB -> **controller execution 397 ms / 18 MiB**
Product with 1229 variants:
   controller execution 450.9 ms / 30 MiB -> **controller execution 338.3 ms / 24 MiB**
Product with 38775 variants:
    controller execution 2006.2 ms / 222 MiB -> **controller execution 372.7 ms / 14 MiB**
Product with 97225 variants:
   Memory limit reached -> **controller execution 358.6 ms / 16 MiB**
         
As you can see, most noticeable performance gains are for products with 10k+ variants.
For products with few variants, there aren't any noticeable changes, for each request execution time fluctuates by +-40ms.
Keep in mind, I'm using local docker environment for mac, so this is not from production.

Also attaching screenshots, for total page load performance. Screenshots are from product with 38775 variants.
Before:
38775-before
After:
38775-after

Hope this helps.

**Comment by jakubtobiasz** [comment]
Thanks for sharing the results. Seems legit to me, can you only take a look at specs, as they're broken now?

**Comment by ulinskas** [comment]
@jakubtobiasz done with specs.

**Review by jakubtobiasz — DISMISSED** [review]
I'll try to gather more reviews this week :). Thanks for your work and improving Sylius' performance 🚀!

File: src/Sylius/Component/Product/spec/Resolver/DefaultProductVariantResolverSpec.php [hunk 1]
```diff
@@ -30,17 +36,25 @@ function it_returns_first_variant(
...
+
         $product->getEnabledVariants()->willReturn($variants);
         $variants->isEmpty()->willReturn(false);
         $variants->first()->willReturn($variant);
 
+
         $this->getVariant($product)->shouldReturn($variant);
     }
 
     function it_returns_null_if_first_variant_is_not_defined(Collection $variants, ProductInterface $product): void
```
Comment (NoResponseMate) [inline src/Sylius/Component/Product/spec/Resolver/DefaultProductVariantResolverSpec.php]: This one should also use the repository, right now it checks nothing.

File: src/Sylius/Component/Product/spec/Resolver/DefaultProductVariantResolverSpec.php [hunk 2]
```diff
@@ -30,17 +36,25 @@ function it_returns_first_variant(
...
     ): void {
+        $product->getId()->willReturn(1);
+        $productVariantRepository->findOneBy([
+            'product' => 1,
+            'enabled' => true,
+        ])->willReturn($variant);
+
         $product->getEnabledVariants()->willReturn($variants);
         $variants->isEmpty()->willReturn(false);
         $variants->first()->willReturn($variant);
```
Comment (NoResponseMate) [inline src/Sylius/Component/Product/spec/Resolver/DefaultProductVariantResolverSpec.php]:
```suggestion
```
No longer used

File: src/Sylius/Component/Product/Resolver/DefaultProductVariantResolver.php [hunk 3]
```diff
@@ -13,17 +13,24 @@
 
 namespace Sylius\Component\Product\Resolver;
 
+use Sylius\Component\Product\Repository\ProductVariantRepositoryInterface;
 use Sylius\Component\Product\Model\ProductInterface;
 use Sylius\Component\Product\Model\ProductVariantInterface;
 
 final class DefaultProductVariantResolver implements ProductVariantResolverInterface
 {
+    public function __construct(private ProductVariantRepositoryInterface $productVariantRepository)
```
Comment (GSadee) [inline src/Sylius/Component/Product/Resolver/DefaultProductVariantResolver.php]: Due to backward compatibility, you should add this argument as optional and leave the old behaviour when this service is not present, and add a deprecation in 1.13 for not using it
```suggestion
    public function __construct(private ?ProductVariantRepositoryInterface $productVariantRepository = null)
```

File: src/Sylius/Component/Product/Resolver/DefaultProductVariantResolver.php [hunk 4]
```diff
@@ -13,17 +13,34 @@
...
+    {
+    }
+
     public function getVariant(ProductInterface $subject): ?ProductVariantInterface
     {
-        if ($subject->getEnabledVariants()->isEmpty()) {
-            return null;
+        if (!$this->productVariantRepository) {
+            trigger_deprecation('sylius/product', '1.12', sprintf('Not passing an instance of %s as constructor argument for %s is deprecated as of Sylius 1.12 and will be mandatory in 1.13.', ProductVariantRepositoryInterface::class, self::class));
+
```
Comment (NoResponseMate) [inline src/Sylius/Component/Product/Resolver/DefaultProductVariantResolver.php]:
```suggestion
```
Deprecations should be added in the next minor

File: src/Sylius/Component/Product/Resolver/DefaultProductVariantResolver.php [hunk 5]
```diff
@@ -13,17 +13,34 @@
...
+                return null;
+            }
+
+            return $subject->getEnabledVariants()->first();
         }
 
-        return $subject->getEnabledVariants()->first();
+        /** @var ProductVariantInterface|null $productVariant */
+        $productVariant = $this->productVariantRepository->findOneBy([
+            'product' => $subject->getId(),
```
Comment (NoResponseMate) [inline src/Sylius/Component/Product/Resolver/DefaultProductVariantResolver.php]: We should return early when a product has no id.

Comment (ulinskas) [reply src/Sylius/Component/Product/Resolver/DefaultProductVariantResolver.php]: I think we should leave this as it is.
You should only use resolver when ID is present, because we are using DB to fetch variant.
If id is optional, than it can lead to confusion.

Example if we do early return:
Product is created in memory and we set variants for his. Everything is still in memory, results aren't flushed to the database.
Method `$product->getVariants()` will return variants, but trying to resolve variants from the default resolver, will result in `null`.

But if we leave it like it is, resolver will throw an error, which is clear indication that something is wrong.

Comment (NoResponseMate) [reply src/Sylius/Component/Product/Resolver/DefaultProductVariantResolver.php]: Even when left as is we're still returning null when the product was not saved, the only difference with an early return would be not triggering doctrine. But now that I think about it, how about falling back to the previous logic in case the product doesn't have an id. The optimization would still be in effect, while the previous behaviour stays the same and we maintain one access point for getting a variant.

Comment (ulinskas) [reply src/Sylius/Component/Product/Resolver/DefaultProductVariantResolver.php]: Agree with you.

File: src/Sylius/Component/Product/spec/Resolver/DefaultProductVariantResolverSpec.php [hunk 6]
```diff
@@ -30,18 +36,28 @@ function it_returns_first_variant(
         ProductInterface $product,
         ProductVariantInterface $variant,
         Collection $variants,
+        ProductVariantRepositoryInterface $productVariantRepository,
```
Comment (Rafikooo) [inline src/Sylius/Component/Product/spec/Resolver/DefaultProductVariantResolverSpec.php:53]: Should be the first argument

File: src/Sylius/Component/Product/spec/Resolver/DefaultProductVariantResolverSpec.php
Comment (Rafikooo) [inline src/Sylius/Component/Product/spec/Resolver/DefaultProductVariantResolverSpec.php:1]: I would be for adding two more specs and update naming f existing ones:
`it_returns_first_variant_if_product_variant_repository_is_(not_)initialized`
and
`it_returns_null_if_first_variant_is_not_defined_and_product_variant_repository_is_(not_)initialized`

wdyt?

**Review by Rafikooo — DISMISSED** [review]

File: src/Sylius/Component/Product/Resolver/DefaultProductVariantResolver.php [hunk 7]
```diff
@@ -13,13 +13,28 @@
...
 namespace Sylius\Component\Product\Resolver;
 
+use Sylius\Component\Product\Repository\ProductVariantRepositoryInterface;
 use Sylius\Component\Product\Model\ProductInterface;
 use Sylius\Component\Product\Model\ProductVariantInterface;
 
 final class DefaultProductVariantResolver implements ProductVariantResolverInterface
 {
+    public function __construct(private ?ProductVariantRepositoryInterface $productVariantRepository = null)
+    {
```
Comment (diimpp) [inline src/Sylius/Component/Product/Resolver/DefaultProductVariantResolver.php:23]: Missing a deprecation as was suggested in https://github.com/Sylius/Sylius/pull/15273#pullrequestreview-1614621056

Here is an [example](https://github.com/Sylius/Sylius/blob/b9ed3125ab5e26a2aa56aed0f743442b70ea8659/src/Sylius/Bundle/AdminBundle/Controller/NotificationController.php#L37-L47)

```php
if (null === $productVariantRepository) {
    trigger_deprecation('sylius/product', '1.13', 'Not passing a service that implements "%s" as a 1st argument of "%s" constructor is deprecated and will be prohibited in 2.0.', ProductVariantRepositoryInterface::class, self::class);
}

File: src/Sylius/Component/Product/Resolver/DefaultProductVariantResolver.php [hunk 8]
```diff
@@ -13,13 +13,28 @@
...
+    }
+
     public function getVariant(ProductInterface $subject): ?ProductVariantInterface
     {
+        if ($this->productVariantRepository && $subject->getId()) {
+            /** @var ProductVariantInterface|null $productVariant */
+            $productVariant = $this->productVariantRepository->findOneBy([
```
Comment (diimpp) [inline src/Sylius/Component/Product/Resolver/DefaultProductVariantResolver.php:30]: I would prefer to have explicit repository methods at component-level and for serious usage, something like `findFirstEnabledForProduct(ProductInterface $product): ?ProductVariantInterface`

WDYT?

**Comment by diimpp** [comment]
Basically every other relation with collection will go out of memory on 1000+ records, by adding optimization at `Component` level we're making assumptions about `Bundle` mappings and `Core`/`Admin` pages/usage.

For example `Order->getLastPayment(?string $state = null): ?PaymentInterface` or `Product->getImagesByType(string $type): Collection` will go out of memory on 1000-9000 records as well.

I would say we should move this change into `Sylius\Bundle\CoreBundle\Product\Resolver\ProductVariantResolver`as decorating service. (Or `Component\Core`, not sure about the difference atm) to keep optimization to "ecommerce-application" level instead of "ecommerce-framework" level, where it's actually needed based on usage.

**tldr**: Optimization should be based on usage and mass-variant usage is allowed at core/admin/api.

**Comment by ulinskas** [comment]
@jakubtobiasz do I miss something else?
Maybe we can merge this PR?

File: src/Sylius/Component/Product/Resolver/DefaultProductVariantResolver.php [hunk 9]
```diff
@@ -13,13 +13,31 @@
...
 
+use Sylius\Component\Product\Repository\ProductVariantRepositoryInterface;
 use Sylius\Component\Product\Model\ProductInterface;
 use Sylius\Component\Product\Model\ProductVariantInterface;
 
 final class DefaultProductVariantResolver implements ProductVariantResolverInterface
 {
+    public function __construct(private ?ProductVariantRepositoryInterface $productVariantRepository = null)
+    {
+        if(!$this->productVariantRepository) {
```
Comment (diimpp) [inline src/Sylius/Component/Product/Resolver/DefaultProductVariantResolver.php]:
```suggestion
        if (!$this->productVariantRepository) {
```
```suggestion
        if (null === $this->productVariantRepository) {
```

**Review by GSadee — DISMISSED** [review]

**Comment by diimpp** [comment]
@GSadee shouldn't this go into 1.13 instead of 1.12 as this is not a regular bugfix? Then deprecation notice can be added immediately. 

Similar performance fix for zone matcher went into 1.13 https://github.com/Sylius/Sylius/pull/15275

**Comment by GSadee** [comment]
> @GSadee shouldn't this go into 1.13 instead of 1.12 as this is not a regular bugfix? Then deprecation notice can be added immediately.
> 
> Similar performance fix for zone matcher went into 1.13 #15275

I have been thinking about this and decided that we could treat it as a bug because of reaching the memory limit with a large number of variants, fix it in this version and add the appropriate deprecation on 1.13. Are you having doubts that we should not consider this a bug?

**Comment by diimpp** [comment]
> Are you having doubts that we should not consider this a bug?

No, personally I don't consider this a bug, every collection relation is susceptible to this issue, try adding 1000 items to a order, 2000 payment attempts for a order, it doesn't make order system broken because of that.

So in my opinion it's not a bugfix for a degraded performance, but rather a feature/improvement to support "bigger" data sets.

**Review by diimpp — APPROVED** [review]

**Review by NoResponseMate — APPROVED** [review]

**Review by TheMilek — APPROVED** [review]

**Comment by ulinskas** [comment]
@jakubtobiasz Friendly reminder that this PR is idle for 2 weeks.

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, Andrius! :tada:

File: src/Sylius/Component/Product/Resolver/DefaultProductVariantResolver.php (hunk 8 above)
Comment (Prometee) [reply src/Sylius/Component/Product/Resolver/DefaultProductVariantResolver.php:30]: Also this query is not using the `position` orderBy product variant field, now the variant selected is not the `0` positioned one but the first created one.
